doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,752 @@
+ * attempt to read a packet from the BIO associated with the given SSL object,
+ * or if there is no associated BIO, or if the SSL object is NULL, then it will
+ * attempt to read a packet from the rbios associated with the SSL_EVENT_CTX and
+ * will then populate the event queue accordingly and fill ev with the first
+ * event. For DTLS this could fulfil the role of DTLSv1_listen() and create a
+ * "new connection" event only when the cookie has been verified.
+ * If after the above the event queue is still empty and SSL_EVENT_TYPE_TICK is
+ * not masked out then it will populate the event queue with any expired timer
+ * events for s if it is non NULL, or any SSL objects otherwise.
+ * If shouldblock is 1 and no events are available this function call will block
+ * until an event is available, or the timeout value has expired. If shouldblock
+ * is 1 and timeout is NULL then the function will block indefinitely until an
+ * event is available. If shouldblock is 0 then then function will always return
+ * immediately.
+ */
+int SSL_EVENT_CTX_get_next_event(SSL_EVENT_CTX *ectx, SSL *s,
+                                 int shouldblock, struct tm *timeout,
+                                 SSL_EVENT *ev);
+
+/*
+ * Add a pending new connection to the new connection queue in the SSL_EVENT_CTX.
+ * This will result in an SSL_EVENT_TYPE_NEW_CONNECTION event being issued. A
+ * subsequent call to SSL_accept will accept the connection and configure the
+ * SSL object for it. Applications should only need to do this for TLS because we
+ * don't otherwise know about the arrival of a new connection. Even for TLS this
+ * is optional (new connections can just be handled as they are now without the
+ * use of SSL_EVENT_CTX if desired). For QUIC and DTLS these new connections
+ * can be detected automatically when SSL_EVENT_CTX_get_next_event() processes a
+ * packet.
+ */
+int SSL_EVENT_CTX_add_new_connection(SSL_EVENT_CTX *ectx, BIO *rbio, BIO *wbio);
+
+/*
+ * Construct and free a new SSL_EVENT object
+ */
+SSL_EVENT *SSL_EVENT_new(void);
+void SSL_EVENT_free(SSL_EVENT *ev);
+
+
+/* Gives an integer value representing the type of event that occurred */
+uint64_t SSL_EVENT_get_event_type(SSL_EVENT *event);
+
+/*
+ * Gets the stream associated with a given event. In the case of "new connection"
+ * this will be NULL. In case the of "new stream" or "connection close" it will
+ * be the "default" stream (i.e. first stream) for the connection
+ */
+SSL *SSL_EVENT_get_stream(SSL_EVENT *event);
+
+/*
+ * Works as normal if the SSL object has already had its BIOs configured.
+ * Otherwise it will check the new connection queue in the SSL_EVENT_CTX. If it
+ * finds a pending new connection then it will "accept" it, adding the BIOs and
+ * running the handshake. In the case of QUIC this will set up the SSL object as
+ * the "default" stream. Will return SSL_ERROR_WANT_READ in the event that no
+ * new conneciton is pending.
+ */
+__owur int SSL_accept(SSL *ssl);
```

Comment:
@hlandau:
> We could add a flags argument here. We don't need to have any flags initially but could have something like NO_DEFAULT_STREAM after the MVP, to create a SSL object representing a QUIC connection without any streams initially.

This is actually an existing function not a new one. I just showed it here so that I could add some commentary about how it is different in the presence of an `SSL_EVENT_CTX`.  Adding a new argument to *this* function isn't possible - although we could have an `SSL_accept_ex()` variant if we want to. Alternatively we also have the existing `SSL_set_options` function which could be used for this purpose.

Reply:
I've added `SSL_CTX_set_params` and `SSL_set_params` for this.